### PR TITLE
Modernize HelloHexagon

### PR DESCRIPTION
**Pull request by abadams**
Updated HelloHexagon to use bin dir for intermediates and to make it a generator. I used a new-style generator with a scheduling lambda. Still trying to feel out the best practice for this. The current options are:

1) Declare all intermediates as class members
2) Capture intermediates in a scheduling lambda
3) Just do the scheduling inside generate() (works for pipelines with no scheduleparams)

### CODE REVIEW

File: apps/HelloHexagon/pipeline.cpp [hunk 1]
```diff
@@ -13,93 +13,90 @@ Expr blur5(Expr x0, Expr x1, Expr x2, Expr x3, Expr x4) {
...
-    std::stringstream hdr;
-    hdr << argv[2] << ".h";
-    blur.compile_to_header(hdr.str(), {input}, argv[2], target);
-    std::stringstream obj;
-    obj << argv[1] << ".o";
-    blur.compile_to_object(obj.str(), {input}, argv[2], target);
 
-    return 0;
-}
+Halide::RegisterGenerator<Blur> register_me{"blur"};
```
Comment (steven-johnson) [inline apps/HelloHexagon/pipeline.cpp]: You can also use the slightly-terser macro:

    HALIDE_REGISTER_GENERATOR(Blur, "blur")

**Comment by steven-johnson** [comment]
IMHO it's probably best to generally avoid (3), if for no other reason than setting a good example. 

I haven't come up with a strong reason to prefer either (1) or (2); my *personal* coding preferences lead me to usually use (1) but that's just me. I suspect that we'll end up with some people strongly preferring (2) and some (1), and at present I'm inclined to just see how that works out -- either it will be confusing or it won't.

**Comment by steven-johnson** [comment]
LGTM

File: apps/HelloHexagon/pipeline.cpp [hunk 2]
```diff
@@ -13,93 +13,90 @@ Expr blur5(Expr x0, Expr x1, Expr x2, Expr x3, Expr x4) {
...
+                blur.dim(0).set_bounds(0, (blur.dim(0).extent()/vector_size)*vector_size);
+                
+                for (int i = 1; i < 3; i++) {
+                    input.dim(i).set_stride((input.dim(i).stride()/vector_size)*vector_size);
+                    blur.dim(i).set_stride((blur.dim(i).stride()/vector_size)*vector_size);
+                }
+            } else {
+                const int vector_size = natural_vector_size<uint8_t>();
+                
+                Func(blur).compute_root()
```
Comment (steven-johnson) [inline apps/HelloHexagon/pipeline.cpp]: The fact that we have to use a cast-to-Func to call schedule methods is a wart I'd like to fix.

One approach I've entertained is using an approach like the HALIDE_BUFFER_FORWARD macros in Buffer.h to make thin forwarders, but so far I've been reluctant to do so, as it feels like there ought to be a more elegant way to solve the problem.

File: apps/HelloHexagon/process.cpp [hunk 3]
```diff
@@ -2,7 +2,7 @@
 #include <memory.h>
 #include <assert.h>
 #include <stdlib.h>
-#include <malloc.h>
+//#include <malloc.h>
```
Comment (ronlieb) [inline apps/HelloHexagon/process.cpp]: remove the commented out include of malloc.h?

**Pull request merged**